NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/clmmate.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "clm mate" 1 "16 May 2014" "clm mate 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
clm mate \- compute best matches between two clusterings

clmmate is not in actual fact a program\&. This manual
page documents the behaviour and options of the clm program when
invoked in mode \fImate\fP\&. The options \fB-h\fP, \fB--apropos\fP,
\fB--version\fP, \fB-set\fP, \fB--nop\fP are accessible
in all \fBclm\fP modes\&. They are described
in the \fBclm\fP manual page\&.
.SH SYNOPSIS

\fBclm mate\fP
\fB[-o\fP fname (\fIoutput file name\fP)\fB]\fP
\fB[-b\fP (\fIomit headers\fP)\fB]\fP
\fB[--one-to-many\fP (\fIrequire multiple hits in <clfile1>\fP)\fB]\fP
\fB[-h\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--apropos\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--version\fP (\fIprint version, exit\fP)\fB]\fP
<clfile1> <clfile2>
.SH DESCRIPTION

\fBclm mate\fP computes for each cluster \fCX\fP in \fCclfile1\fP all clusters
\fCY\fP in \fCclfile2\fP that have non-empty intersection and outputs
a line with the data points listed below\&.

.di ZV
.in 0
.nf \fC
   overlap(X,Y)               # 2 * size(meet(X,Y)) / (size(X)+size(Y))
   index(X)                   # name of cluster
   index(Y)                   # name of cluster
   size(meet(X,Y))
   size(X-Y)                  # size of left difference
   size(Y-X)                  # size of right difference
   size(X)
   size(Y)
   projection(X, clfile2)     # see below
   projection(Y, clfile1)     # see below
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

The projected size of a cluster \fCX\fP relative to a clustering \fCK\fP is
simply the sum of all the nodes shared between any cluster \fCY\fP in \fCK\fP
and \fCX\fP, duplications allowed\&. For example, the projected size of
\fC(0,1)\fP relative to \fC{(0,2,4), (1,4,9), (1,3,5)}\fP equals \fC3\fP\&.

The overlap between \fCX\fP and \fCY\fP is exactly
1\&.0 if the two clusters are identical, and for nearly identical
clusterings the score will be close to 1\&.0\&.

All of this information can also be obtained from the
contingency matrix defined for two clusterings\&.
The \fC[i,j]\fP row-column entry in a contigency matrix between
to clusterings gives the number of entries in the intersection
between cluster\ \&\fCi\fP and cluster\ \&\fCj\fP from the respective
clusterings\&. The other information is implicitly present;
the total number of nodes in clusters\ \&\fCi\fP and\ \&\fCj\fP
for example can be obtained as the sum of entries in row\ \&\fCi\fP
and column\ \&\fCj\fP respectively, and the difference counts
can then be obtained by substracting the intersection count\&.
The contingency matrix can easily be computed using \fBmcx\fP;
e\&.g\&.

.di ZV
.in 0
.nf \fC

mcx /clfile2 lm /clfile1 lm tp mul /ting wm
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

will create the contingency matrix in mcl matrix format
in the file \fCting\fP, where columns range over the clusters
in \fCclfile1\fP\&.

The output can be put to good use by sorting it numerically on
that first score field\&. It is advisable to use a stable sort routine
(use the \fB-s\fP option for UNIX sort)
From this information one can quickly extract the closest
clusters between two clusterings\&.
.SH OPTIONS

.ZI 2m "\fB-o\fP fname (\fIoutput file name\fP)"
\&
.br
Specify the name of the output file\&.
.in -2m

.ZI 2m "\fB-b\fP (\fIomit headers\fP)"
\&
.br
Batch mode, omit column names\&.
.in -2m

.ZI 2m "\fB--one-to-many\fP (\fIrequire multiple hits in <clfile1>\fP)"
\&
.br
Do not output information for clusters in the first file
that are subset of a cluster in the second file\&.
.in -2m
.SH AUTHOR

Stijn van Dongen\&.
.SH SEE ALSO

\fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.


NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shcl/clmmate.c
/*   (C) Copyright 2003, 2004, 2005 Stijn van Dongen
 *   (C) Copyright 2006, 2007 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/

/* TODO
 * compute when inner loop can be broken; all of xproj has been accounted for.
 *    (sum meetsize)
 * ? adapt or warn on mismatched domains.
*/


#include <string.h>
#include <stdio.h>

#include "clm.h"
#include "clmmate.h"

#include "util/types.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/compile.h"

#include "impala/matrix.h"
#include "impala/io.h"
#include "impala/iface.h"
#include "impala/compose.h"
#include "impala/ivp.h"
#include "impala/app.h"

#include "clew/clm.h"

static const char* me = "clmmate";

enum
{  MY_OPT_OUTPUT = CLM_DISP_UNUSED
,  MY_OPT_BATCH
,  MY_OPT_1TM
}  ;


static mcxOptAnchor mateOptions[] =
{  {  "--one-to-many"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_1TM
   ,  NULL
   ,  "only output pairs with multiple incidences in left <cl file>"
   }
,  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_OUTPUT
   ,  "<fname>"
   ,  "output file name"
   }
,  {  "-b"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_BATCH
   ,  NULL
   ,  "batch mode, omit leading headers"
   }
,  {  NULL ,  0 ,  0 ,  NULL, NULL}
}  ;


static mcxbool legend   =  -1;
static mcxIO*  xfout    =  (void*) -1;
static mcxbool one2many =  -1;


static mcxstatus mateInit
(  void
)
   {  xfout = mcxIOnew("-", "w")
   ;  legend = TRUE
   ;  one2many = FALSE
   ;  return STATUS_OK
;  }


static mcxstatus mateArgHandle
(  int optid
,  const char* val
)
   {  switch(optid)
      {  case MY_OPT_1TM
      :  one2many = TRUE
      ;  break
      ;

         case MY_OPT_BATCH
      :  legend = FALSE
      ;  break
      ;

         case MY_OPT_OUTPUT
      :  mcxIOnewName(xfout, val)
      ;  break
      ;

         default
      :  return STATUS_FAIL
      ;
      }
      return STATUS_OK
;  }


static mcxstatus mateMain
(  int         argc_unused    cpl__unused
,  const char* argv[]
)
   {  mcxIO* xfx, *xfy
   ;  mclx* mx, *my, *meet, *teem, *myt
   ;  dim x, y

   ;  mcxIOopen(xfout, EXIT_ON_FAIL)

   ;  xfx =  mcxIOnew(argv[0], "r")
   ;  mx  =  mclxRead(xfx, EXIT_ON_FAIL)
   ;  mcxIOclose(xfx)
   ;  xfy =  mcxIOnew(argv[1], "r")
   ;  my  =  mclxRead(xfy, EXIT_ON_FAIL)
   ;  myt =  mclxTranspose(my)

   ;  if (!MCLD_EQUAL(mx->dom_rows, my->dom_rows))
      mcxDie(1, me, "domains are not equal")

   ;  meet=  mclxCompose(myt, mx, 0, 0)      /* fixme thread interface */
   ;  teem=  mclxTranspose(meet)

   ;  if (legend)
      fprintf
      (  xfout->fp
      ,  "%-10s %6s %6s %6s %6s %6s %6s %6s\n"
      ,  "overlap"
      ,  "x-idx"
      ,  "y-idx"
      ,  "meet"
      ,  "xdiff"
      ,  "ydiff"
      ,  "x-size"
      ,  "y-size"
      )

   ;  for (x=0;x<N_COLS(meet);x++)
      {  mclv* xvec = meet->cols+x
      ;  long X = xvec->vid
      ;  long xsize = mx->cols[x].n_ivps

      ;  if (one2many && xvec->n_ivps < 2)
         continue

      ;  for (y=0;y<N_COLS(teem);y++)
         {  mclv* yvec = teem->cols+y
         ;  long Y = yvec->vid
         ;  long ysize = my->cols[y].n_ivps
         ;  double twinfac
         ;  long meetsize
         ;  mclp* ivp = mclvGetIvp(yvec, X, NULL)
         ;  if (!ivp)
            continue

         /*
          * meet size, left diff, right diff, right size.
         */

         ;  meetsize = ivp->val

         ;  if (!xsize && !ysize)         /* paranoia */
            continue

         ;  twinfac = 2 * meetsize / ( (double) (xsize + ysize) )

         ;  if (xfout)
            fprintf
            (  xfout->fp
            ,  "%-10.3f %6ld %6ld %6ld %6ld %6ld %6ld %6ld\n"
            ,  twinfac
            ,  X
            ,  Y
            ,  meetsize
            ,  xsize - meetsize
            ,  ysize - meetsize
            ,  xsize
            ,  ysize
            )
      ;  }
      }
      return STATUS_OK
;  }


mcxDispHook* mcxDispHookMate
(  void
)
   {  static mcxDispHook mateEntry
   =  {  "mate"
      ,  "mate [options] <cl file> <cl file>"
      ,  mateOptions
      ,  sizeof(mateOptions)/sizeof(mcxOptAnchor) - 1
      ,  mateArgHandle
      ,  mateInit
      ,  mateMain
      ,  2
      ,  2
      ,  MCX_DISP_MANUAL
      }
   ;  return &mateEntry
;  }


